pkg/operator/controller/ingress/status.go

```diff
@@ -171,35 +181,109 @@ func computeDeploymentPodsScheduledCondition(deployment *appsv1.Deployment, pods
-	for _, cond := range deployment.Status.Conditions {
-		if cond.Type != appsv1.DeploymentAvailable {
-			continue
+// by inspecting the following:
+// 1) the Available condition of deployment
+// 2) the DNSReady condition
+// 3) the LoadBalancerReady condition
+// The ingresscontroller is judged Available only if all 3 conditions are true
+func computeIngressAvailableCondition(conditions []operatorv1.OperatorCondition) []operatorv1.OperatorCondition {
+	conditionsMap := make(map[string]*operatorv1.OperatorCondition)
+	for i := range conditions {
+		conditionsMap[conditions[i].Type] = &conditions[i]
+	}
+	expected := []expectedCondition{
+		{
+			condition: string(appsv1.DeploymentAvailable),
+			status:    operatorv1.ConditionTrue,
+			mustHave:  true,
+		},
+		{
+			condition: operatorv1.DNSReadyIngressConditionType,
+			status:    operatorv1.ConditionTrue,
+			ifConditionsTrue: []string{
+				operatorv1.LoadBalancerManagedIngressConditionType,
+				operatorv1.LoadBalancerReadyIngressConditionType,
+				operatorv1.DNSManagedIngressConditionType,
+			},
+			gracePeriod: time.Second * 30,
+			mustHave:    true,
+		},
+		{
+			condition:        operatorv1.LoadBalancerReadyIngressConditionType,
+			status:           operatorv1.ConditionTrue,
+			ifConditionsTrue: []string{operatorv1.LoadBalancerManagedIngressConditionType},
+			gracePeriod:      time.Second * 90,
+			mustHave:         true,
+		},
+	}
+	_, unavailableConditions := checkAnnotatedConditions(expected, conditionsMap)
+
+	// TODO - check with team - do we want to use a grace period when checking Available?
+
+	if len(unavailableConditions) != 0 {
+		degraded := formatConditions(unavailableConditions)
+		return []operatorv1.OperatorCondition{
+			{
+				Type:    operatorv1.IngressControllerAvailableConditionType,
+				Status:  operatorv1.ConditionFalse,
+				Reason:  "IngressControllerUnavailable",
+				Message: "One or more status conditions indicate unavailable: " + degraded,
+			},
 		}
-		switch cond.Status {
-		case corev1.ConditionTrue:
-			return operatorv1.OperatorCondition{
+	} else {
+		return []operatorv1.OperatorCondition{
+			{
 				Type:   operatorv1.IngressControllerAvailableConditionType,
 				Status: operatorv1.ConditionTrue,
-			}
-		case corev1.ConditionFalse:
-			return operatorv1.OperatorCondition{
-				Type:    operatorv1.IngressControllerAvailableConditionType,
-				Status:  operatorv1.ConditionFalse,
-				Reason:  cond.Reason,
-				Message: "The deployment is unavailable: " + cond.Message,
-			}
+			},
 		}
 	}
+}
 
-	return operatorv1.OperatorCondition{
-		Type:    operatorv1.IngressControllerAvailableConditionType,
-		Status:  operatorv1.ConditionFalse,
-		Reason:  "DeploymentAvailabilityUnknown",
-		Message: "The deployment's Available condition couldn't be interpreted",
+func checkAnnotatedConditions(expecteds []expectedCondition, conditionsMap map[string]*operatorv1.OperatorCondition) ([]*operatorv1.OperatorCondition, []*operatorv1.OperatorCondition) {
```

Comment:
Could this be named simply `checkAnnotatedConditions`?

Reply:
Not sure what you mean?  It is already named checkAnnotatedConditions...

Reply:
Sorry, I meant to suggest `checkConditions`.

Reply:
Ok, I will change it

Reply:
Done.